fish-rust/src/common.rs

```diff
@@ -1752,72 +1742,34 @@ pub type Cleanup<T, F> = ScopeGuard<T, F>;
 ///
 /// // hello will be written first, then goodbye.
 /// ```
-pub struct ScopeGuard<T, F: FnOnce(&mut T)> {
-    captured: ManuallyDrop<T>,
-    on_drop: Option<F>,
-}
+pub struct ScopeGuard<T, F: FnOnce(T)>(Option<(T, F)>);
 
-impl<T, F> ScopeGuard<T, F>
-where
-    F: FnOnce(&mut T),
-{
+impl<T, F: FnOnce(T)> ScopeGuard<T, F> {
     /// Creates a new `ScopeGuard` wrapping `value`. The `on_drop` callback is executed when the
     /// ScopeGuard's lifetime expires or when it is manually dropped.
     pub fn new(value: T, on_drop: F) -> Self {
-        Self {
-            captured: ManuallyDrop::new(value),
-            on_drop: Some(on_drop),
-        }
-    }
-
-    /// Cancel the unwind operation, e.g. do not call the previously passed-in `on_drop` callback
-    /// when the current scope expires.
-    pub fn cancel(guard: &mut Self) {
-        guard.on_drop.take();
-    }
-
-    /// Cancels the unwind operation like [`ScopeGuard::cancel()`] but also returns the captured
-    /// value (consuming the `ScopeGuard` in the process).
-    pub fn rollback(mut guard: Self) -> T {
```

Comment:
Elsewhere I've used the C++ stdlib's equivalent of `cancel()` like this:

```rust
let mut populate_error = ScopeGuard::new(populate_error_fn);
if error-scenario-1 {
   return Err(());
}
if error-scenario-2 {
   return Err(());
}
populate_error.cancel();
Ok(())
```

I don't expect this to be a common pattern but it's probably useful in some situations.

I wonder if cancel() and rollback() actually need `ManuallyDrop`?

Reply:
That's actually a very interesting approach, though I'll be honest and say it would take me a second to understand what was being done. It's funnily enough a very timely example of what I said above ("because I had a feeling it would either be required or just that its presence would lead one to discover a better/easier way of going about accomplishing a simple task")

Reply:
Yeah I wouldn't have discovered this if it wasn't in the stdlib but I'm not sure if it's actually useful in Rust.